### ISSUE TITLE
`null_check_on_nullable_type_parameter` should work on all pattern assignments

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#4218]
Records work but we should ensure that objects, lists and maps are covered too.

```dart
class A<T> {
  T? a;
  A(this.a);
}

f<T>(T? t, T? u) {
  var A(a: t!) = A(u);
}
```

```dart
void f<T>(List<T?> l){
  var [x!, y] = l;
}
```

```dart
void f<T>(Map<String, T?> m){
  var {'x': y!} = m;
}
```

**Comment by pq** [comment]
@srawlins, @bwilkerson, does this seem like a valid test for the object case?


```dart
class A<T> {
  T? a;
  A(this.a);
}

f<T>(T? t, T? u) {
  var A(a: t!) = A(u);
}
```

**Comment by bwilkerson** [comment]
I'm not sure because of the complexity of type inference.

I think we could simplify it by removing the type parameter on the class and have the same effect without the confusion of (a) using the same name for two different type parameters and (b) having the type arguments for the pattern and constructor invocation be inferred.

We could also get rid of the constructor invocation by passing in an instance of `A` rather than a value to pass to the constructor.
```
class A {
  Object? a;
  A(this.a);
}

void f<T>(T? t, A u) {
  var A(a: t!) = u;
}
```

With those simplifications, I think it's a valid test. I just don't know whether the type inference of the type arguments made the previous form valid or invalid.

**Comment by srawlins** [comment]
I think Brian's feedback is right.

**Comment by pq** [comment]
Super. Thank you both!
